securitas/templates/_form_macros.html

```diff
@@ -1,24 +1,39 @@
 {% macro with_errors(field) %}
-  {% if kwargs.pop('nolabel', '') %}
-    {% set label = False %}
-  {% else %}
-    {% set label = True %}
-  {% endif %}
-  {% if field.errors %}
-    {% set css_class = 'form-control is-invalid ' + kwargs.pop('class', '') %}
-  {% else %}
-    {% set css_class = 'form-control ' + kwargs.pop('class', '') %}
-  {% endif %}
+  {% set css_class = 'form-control ' + kwargs.pop('class', '') %}
 
-  {% if label %}
-    <label for="{{ field.name }}" class="mb-0 font-weight-bold">{{ field.label }}</label>
-  {% endif %}
-  {{ field(class=css_class, **kwargs) }}
-  
-  {% if field.errors %}
-  <div class="invalid-feedback">
-      {{ field.errors|map('escape')|join(' ') }}
-  </div>
+  {% if field.type == 'FieldList' %}
+    {# append 4 empty entries at the bottom of the fieldlist #}
+    {% for i in range(4) %}
+      {% set entry = field.append_entry() %}
+    {% endfor %}
+
+    {% if kwargs.pop('label', True) %}
+      <label for="{{ field.name }}" class="mb-0 font-weight-bold">{{ field[0].label }}</label>
+    {% endif %}
+    {% for subfield in field %}
+      {% if subfield.errors %}
+        {% set css_class = css_class + ' is-invalid ' %}
+      {% endif %}
+      {{ subfield(class=css_class, **kwargs) }}
+      {% if subfield.errors %}
+        <div class="invalid-feedback">
+          {{ subfield.errors|map('escape')|join(' ') }}
+        </div>
+      {% endif %}
+    {% endfor %}
+  {% else %}
+    {% if field.errors %}
+      {% set css_class = css_class + ' is-invalid ' %}
+    {% endif %}
+    {% if kwargs.pop('label', True) %}
+      <label for="{{ field.name }}" class="mb-0 font-weight-bold">{{ field.label }}</label>
+    {% endif %}
+    {{ field(class=css_class, **kwargs) }}
+    {% if field.errors %}
+    <div class="invalid-feedback">
+        {{ field.errors|map('escape')|join(' ') }}
+    </div>
```

Comment:
I think we can factor some of this code, because I can see some duplication. I suggest the following change:
```diff
--- a/securitas/templates/_form_macros.html
+++ b/securitas/templates/_form_macros.html
@@ -1,6 +1,21 @@
-{% macro with_errors(field) %}
+{% macro form_field(field) %}
   {% set css_class = 'form-control ' + kwargs.pop('class', '') %}
+  {% if field.errors %}
+    {% set css_class = css_class + ' is-invalid ' %}
+  {% endif %}
+  {% if kwargs.pop('label', True) %}
+    <label for="{{ field.name }}" class="mb-0 font-weight-bold">{{ field.label }}</label>
+  {% endif %}
+  {{ field(class=css_class, **kwargs) }}
+  {% if field.errors %}
+  <div class="invalid-feedback">
+      {{ field.errors|map('escape')|join(' ') }}
+  </div>
+  {% endif %}
+{% endmacro %}
+
 
+{% macro with_errors(field) %}
   {% if field.type == 'FieldList' %}
     {# append 4 empty entries at the bottom of the fieldlist #}
     {% for i in range(4) %}
@@ -11,32 +26,14 @@
       <label for="{{ field.name }}" class="mb-0 font-weight-bold">{{ field[0].label }}</label>
     {% endif %}
     {% for subfield in field %}
-      {% if subfield.errors %}
-        {% set css_class = css_class + ' is-invalid ' %}
-      {% endif %}
-      {{ subfield(class=css_class, **kwargs) }}
-      {% if subfield.errors %}
-        <div class="invalid-feedback">
-          {{ subfield.errors|map('escape')|join(' ') }}
-        </div>
-      {% endif %}
+      {{ form_field(subfield, label=False, **kwargs) }}
     {% endfor %}
   {% else %}
-    {% if field.errors %}
-      {% set css_class = css_class + ' is-invalid ' %}
-    {% endif %}
-    {% if kwargs.pop('label', True) %}
-      <label for="{{ field.name }}" class="mb-0 font-weight-bold">{{ field.label }}</label>
-    {% endif %}
-    {{ field(class=css_class, **kwargs) }}
-    {% if field.errors %}
-    <div class="invalid-feedback">
-        {{ field.errors|map('escape')|join(' ') }}
-    </div>
-    {% endif %}
+    {{ form_field(field, **kwargs) }}
   {% endif %}
 {% endmacro %}
 
+
 {% macro non_field_errors(form) %}
   {% if form.errors.non_field_errors %}
     <div id="formerrors">
```

What do you think?

Reply:
+1 and updatrred